### PR TITLE
[ESP32] Notify-update-applied command and added ota initialization to common space

### LINKED ISSUE #13760: [ESP32] NotifyUpdateApplied should also be called once the device has booted to a new version
#### Problem
#13555 has added support for `NotifyUpdateApplied`. #13620  will restart ESP32 device automatically after applying an OTA image.

In the same flow NotifyUpdateApplied should also be called once device boots into a new firmware version
 #### Proposed Solution
- Add support for sending `NotifyUpdateApplied` command

### PR DESCRIPTION
**Pull request by jadhavrohit924**
#### Problem
* Added IsFirstImageRun and ConfirmCurrentImage definations for ESP32.
* Fixed initialization order of ESP32.
* Fixes #13760 

#### Change overview
Schedule notify-update-applied command.

#### Testing
* Manually tested esp32 apps.

### CODE REVIEW

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp [hunk 1]
```diff
@@ -48,6 +48,7 @@ using namespace app::Clusters::OtaSoftwareUpdateRequestor::Structs;
 constexpr uint32_t kDelayQueryUponCommissioningSec = 30; // Delay before sending the initial image query after commissioning
 constexpr uint32_t kImmediateStartDelaySec         = 1;  // Delay before sending a query in response to UrgentUpdateAvailable
 constexpr System::Clock::Seconds32 kDefaultDelayedActionTime = System::Clock::Seconds32(120);
+constexpr uint32_t kDelayNotifyUpdateApplied                 = 10; // Delay before sending notify update applied command
```
Comment (Damian-Nordic) [inline src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp:51]: Can this be made configurable? I think in most cases the delay is not really needed and, so a shorter delay will speed up smoke-testing.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp [hunk 2]
```diff
@@ -74,7 +84,8 @@ void GenericOTARequestorDriver::Init(OTARequestorInterface * requestor, OTAImage
                 return;
             }
 
-            mRequestor->NotifyUpdateApplied();
+            chip::DeviceLayer::SystemLayer().StartTimer(chip::System::Clock::Seconds32(kDelayNotifyUpdateApplied),
```
[the remark this answers is not in the document]
Comment (jadhavrohit924) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp]: Done

**Review by selissia — APPROVED** [review]

**Review by Damian-Nordic — APPROVED** [review]

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp [hunk 3]
```diff
@@ -64,18 +71,20 @@ void GenericOTARequestorDriver::Init(OTARequestorInterface * requestor, OTAImage
...
-
-            mRequestor->NotifyUpdateApplied();
-        });
+        ScheduleDelayedAction(
+            mDelayNotifyUpdateAppliedSeconds,
+            [](System::Layer *, void * context) {
+                CHIP_ERROR error = ToDriver(context)->mImageProcessor->ConfirmCurrentImage();
+                if (error != CHIP_NO_ERROR)
+                {
+                    ChipLogError(SoftwareUpdate, "Failed to confirm image: %" CHIP_ERROR_FORMAT, error.Format());
```
Comment (isiu-apple) [inline src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp:80]: If we fail to confirm current image, will we implement rolling back to previous working version?

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp [hunk 4]
```diff
@@ -64,18 +71,20 @@ void GenericOTARequestorDriver::Init(OTARequestorInterface * requestor, OTAImage
...
-            {
-                ChipLogError(SoftwareUpdate, "Failed to confirm image: %" CHIP_ERROR_FORMAT, error.Format());
-                mRequestor->Reset();
-                return;
-            }
-
-            mRequestor->NotifyUpdateApplied();
-        });
+        ScheduleDelayedAction(
+            mDelayNotifyUpdateAppliedSeconds,
```
Comment (isiu-apple) [inline src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp:75]: Looks like mDelayNotifyUpdateAppliedSeconds is used to delay confirming current image, which in turns causes NotifyUpdateApplied() to be delayed.  Should we either rename this to mDelayConfirmCurrentImageSec or have mDelayNotifyUpdateAppliedSeconds be applied directly to the delay of calling NotifyUpdateApplied()?

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp (hunk 1 above)
Comment (isiu-apple) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp]: nit: kDelayNotifyUpdateApplied -> kDelayNotifyUpdateAppliedSec

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp (hunk 3 above)
Comment (carol-apple) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp:80]: Not for v1. It was discussed that reverting back to a previous good version can be done implicitly by the bootloader.

File: src/platform/ESP32/OTAImageProcessorImpl.cpp [hunk 5]
```diff
@@ -38,6 +39,35 @@ void HandleRestart(Layer * systemLayer, void * appState)
...
+CHIP_ERROR OTAImageProcessorImpl::ConfirmCurrentImage()
+{
+    OTARequestorInterface * requestor = chip::GetRequestorInstance();
+    if (requestor == nullptr)
+    {
+        return CHIP_ERROR_INTERNAL;
+    }
+
+    uint32_t currentVersion;
+    ReturnErrorOnFailure(DeviceLayer::ConfigurationMgr().GetSoftwareVersion(currentVersion));
```
Comment (carol-apple) [inline src/platform/ESP32/OTAImageProcessorImpl.cpp:62]: FYI, this was a simple implementation done for Darwin/Linux platforms to check the image. I'm not sure if ESP32 following that as an example is necessarily the right thing to do?

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp [hunk 6]
```diff
@@ -48,6 +48,8 @@ using namespace app::Clusters::OtaSoftwareUpdateRequestor::Structs;
 constexpr uint32_t kDelayQueryUponCommissioningSec = 30; // Delay before sending the initial image query after commissioning
 constexpr uint32_t kImmediateStartDelaySec         = 1;  // Delay before sending a query in response to UrgentUpdateAvailable
 constexpr System::Clock::Seconds32 kDefaultDelayedActionTime = System::Clock::Seconds32(120);
+System::Clock::Seconds32 mDelayNotifyUpdateAppliedSeconds =
```
Comment (carol-apple) [inline src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp:51]: Why is there a delay to notify? This isn't part of the spec. Is there a particular reason ESP32 needs to have a delay prior to notifying?

Comment (jadhavrohit924) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp:51]: Yes, after reboot wifi takes some time to start.

Comment (shubhamdp) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp:51]: This routine is executed very early and Wi-Fi is not yet connected at that point, so buying some time till the wifi connects.

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp (hunk 4 above)
Comment (jadhavrohit924) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp]: Remaned to mDelayConfirmCurrentImageSec.

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp [hunk 7]
```diff
@@ -56,6 +58,11 @@ GenericOTARequestorDriver * ToDriver(void * context)
 
 } // namespace
 
+void GenericOTARequestorDriver::DelayNotifyUpdateAppliedAction(uint32_t seconds)
```
[the remark this answers is not in the document]
Comment (jadhavrohit924) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp]: Renamed to SetDelayConfirmCurrentImageSec().

File: src/platform/ESP32/OTAImageProcessorImpl.cpp (hunk 5 above)
Comment (jadhavrohit924) [reply src/platform/ESP32/OTAImageProcessorImpl.cpp:62]: ESP32 don’t have platform specific API's.

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp [hunk 8]
```diff
@@ -48,6 +48,8 @@ using namespace app::Clusters::OtaSoftwareUpdateRequestor::Structs;
 constexpr uint32_t kDelayQueryUponCommissioningSec = 30; // Delay before sending the initial image query after commissioning
 constexpr uint32_t kImmediateStartDelaySec         = 1;  // Delay before sending a query in response to UrgentUpdateAvailable
 constexpr System::Clock::Seconds32 kDefaultDelayedActionTime = System::Clock::Seconds32(120);
+System::Clock::Seconds32 mDelayConfirmCurrentImageSec =
+    System::Clock::Seconds32(0); // Delay before sending notify update applied command
```
Comment (isiu-apple) [inline src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp:52]: Can we update the comment to match the variable as well? e.g., Delay before confirming current image (in seconds)

Comment (jadhavrohit924) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp]: Sure, done

File: examples/all-clusters-app/esp32/main/main.cpp [hunk 9]
```diff
@@ -70,7 +70,8 @@ using namespace ::chip::DeviceLayer;
 // Used to indicate that an IP address has been added to the QRCode
 #define EXAMPLE_VENDOR_TAG_IP 1
 
-const char * TAG = "all-clusters-app";
+const char * TAG                  = "all-clusters-app";
+const uint32_t delayNotifySeconds = 10;
```
[the remark this answers is not in the document]
Comment (jadhavrohit924) [reply examples/all-clusters-app/esp32/main/main.cpp]: Sure, done

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp (hunk 6 above)
Comment (carol-apple) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp]: Wondering if this delay should be done at the application level then?

Comment (carol-apple) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp]: I'm wondering if `GenericOTARequestorDriver::Init` should just be delayed from the application level instead?

Comment (carol-apple) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp]: Yes, I think that's probably the right approach - create some kind of callback for post-connection to network. The notify will be automatically done in `DefaultOTARequestorDriver::Init` if on first boot from an upgrade anyway so perhaps the callback (indicating W-Fi is connected) is the one responsible for calling all the OTA related initialization code? Even in non-upgrade scenario, we probably should just wait for Wi-Fi to be connected before all the inits are called, right?

Comment (jadhavrohit924) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp]: @carol-apple @dhrishi accordingly made changes please review.

File: examples/all-clusters-app/esp32/main/DeviceCallbacks.cpp [hunk 10]
```diff
@@ -55,7 +56,8 @@ using namespace ::chip::System;
 using namespace ::chip::DeviceLayer;
 using namespace chip::app;
 
-constexpr uint32_t kIdentifyTimerDelayMS = 250;
+constexpr uint32_t kIdentifyTimerDelayMS     = 250;
+constexpr uint32_t kInitOTARequestorDelaySec = 3;
```
Comment (carol-apple) [inline examples/all-clusters-app/esp32/main/DeviceCallbacks.cpp:60]: Is 3 seconds sufficient amount of time for WiFi to start?

**Review by carol-apple — APPROVED** [review]

**Review by dhrishi — APPROVED** [review]

File: examples/all-clusters-app/esp32/main/DeviceCallbacks.cpp (hunk 10 above)
Comment (jadhavrohit924) [reply examples/all-clusters-app/esp32/main/DeviceCallbacks.cpp:60]: Wifi is already started, also got ip. After 3s we are initialising OTA.

**Review by shubhamdp — APPROVED** [review]

**Pull request merged**